core/shared/src/main/scala/zio/ZRef.scala

```diff
@@ -0,0 +1,278 @@
+   * @tparam EC type of the error value of the `ZRef` to be modified
+   * @return `ZRef[EC, EB, A, B]` a new `ZRef` with the modified error value
+   */
+  final def mapError[EC](f: EA => EC): ZRef[EC, EB, A, B] = new ZRef[EC, EB, A, B] {
+    def set(a: A): IO[EC, Unit]          = self.set(a).mapError(f)
+    def get: IO[EB, B]                   = self.get
+    def setAsync(a: A): zio.IO[EC, Unit] = self.setAsync(a).mapError(f)
+  }
+
+  /**
+   * Atomically modifies the `ZRef` with the specified function, which computes
+   * a return value for the modification. This is a more powerful version of
+   * `update`.
+   *
+   * @param f function which computes a return value for the modification
+   * @tparam C type of the value of the `ZRef` to be modified
+   * @tparam E type of the error of the `ZRef` to be modified
+   * @return `IO[E, C]` modified value the `ZRef`
+   */
+  def modify[E, C](f: B => (C, A))(implicit ea: EA <:< E, eb: EB <:< E): IO[E, C] = {
+    val unified = unifyError(ea, eb)
+    unified.get.flatMap { b =>
+      val (c, a) = f(b)
+      unified.set(a).as(c)
+    }.uninterruptible
+  }
+
+  /**
+   * Atomically modifies the `ZRef` with the specified partial function, which computes
+   * a return value for the modification if the function is defined in the current value
+   * otherwise it returns a default value.
+   * This is a more powerful version of `updateSome`.
+   *
+   * @param default default value to be returned if the partial function is not defined on the current value
+   * @param pf partial function to be computed on the current value if defined on the current value
+   * @tparam C type of the value of the `ZRef` to be modified
+   * @return `IO[E,C]` modified value of the `ZRef`
+   */
+  def modifySome[E, C](default: C)(pf: PartialFunction[B, (C, A)])(implicit ea: EA <:< E, eb: EB <:< E): IO[E, C] = {
+    val unified = unifyError(ea, eb)
+    def f: B => IO[E, C] = { b =>
+      pf.lift(b).fold(unified.as(default).get) { case (c, a) => unified.set(a).as(c) }
+    }
+    unified.get.flatMap[Any, E, C] { f }.uninterruptible
+  }
+
+  /**
+   * Discard the content of the ZRef and replace it with `Unit`
+   *
+   * @return `ZRef[EA, EB, A, Unit]` a new `ZRef` with the discarded value
+   */
+  final def unit: ZRef[EA, EB, A, Unit] = as(())
+
+  private[zio] def unifyError[E](ea: EA => E, eb: EB => E): ZRef[E, E, A, B] =
+    new ZRef[E, E, A, B] {
+      def set(a: A): IO[E, Unit]          = self.set(a).mapError(ea)
+      def setAsync(a: A): zio.IO[E, Unit] = self.setAsync(a).mapError(ea)
+      def get: IO[E, B]                   = self.get.mapError(eb)
+    }
+
+  /**
+   * Atomically modifies the `ZRef` with the specified function. This is not
+   *
+   * @param f function to atomically modify the `ZRef`
+   * @tparam E type of the unified error value of the `ZRef`
+   * @return `IO[E,B]` modified value of the `ZRef`
+   */
+  def update[E](f: B => A)(implicit ea: EA <:< E, eb: EB <:< E): IO[E, B] = {
```

Comment:
How good is type inference here?

Reply:
you can check out the tests, it works fine, it infers correctly.
The issue now is `modify` and `modifySome` because of the new `E` type param, it needs to be specified and leads to this atrocity 
https://github.com/zio/zio/pull/1198/commits/e7be76639dcceeec4287a31fb00d6ff7e2e65249#diff-e2c16046e05624c39e46b9e4c2870b9cR2138
anything that can be done to make it infer the `E` ?